### PR TITLE
Morello exception handling cleanup/fixes and boot isb

### PR DESCRIPTION
**Pull request by jrtc27**
Should supersede #921.

### CODE REVIEW

File: sys/arm64/arm64/exception.S [hunk 1]
```diff
@@ -38,32 +38,39 @@ __FBSDID("$FreeBSD$");
...
-	stp	x14, x15, [sp, #(TF_X + 14 * 8)]
-	stp	x12, x13, [sp, #(TF_X + 12 * 8)]
-	stp	x10, x11, [sp, #(TF_X + 10 * 8)]
-	stp	x8,  x9,  [sp, #(TF_X + 8  * 8)]
-	stp	x6,  x7,  [sp, #(TF_X + 6  * 8)]
-	stp	x4,  x5,  [sp, #(TF_X + 4  * 8)]
-	stp	x2,  x3,  [sp, #(TF_X + 2  * 8)]
-	stp	x0,  x1,  [sp, #(TF_X + 0  * 8)]
-	mrs	x10, elr_el1
+	stp	x29, x30, [sp, #-16]!
```
Comment (arichardson) [inline sys/arm64/arm64/exception.S:41]: I guess this diff is needed due to the immediate range in hybrid?

File: sys/arm64/arm64/exception.S [hunk 2]
```diff
@@ -129,20 +129,26 @@ __FBSDID("$FreeBSD$");
 	ldp	CAP(12), CAP(13), [sp, #(TF_X + 12 * CAP_WIDTH)]
 	ldp	CAP(14), CAP(15), [sp, #(TF_X + 14 * CAP_WIDTH)]
 	ldp	CAP(16), CAP(17), [sp, #(TF_X + 16 * CAP_WIDTH)]
+#if !__has_feature(capabilities) || defined(__CHERI_PURE_CAPABILITY__)
```
Comment (arichardson) [inline sys/arm64/arm64/exception.S:132]: If we see more instances of this ifdef we might want to add a new
`__has_feature(cheri_hybrid)` or `__CHERI_HYBRID_CAPABILITY__`

**Review by arichardson — APPROVED** [review]

File: sys/arm64/arm64/exception.S (hunk 1 above)
Comment (jrtc27) [reply sys/arm64/arm64/exception.S:41]: Yes, see the relevant commit message

File: sys/arm64/arm64/exception.S (hunk 2 above)
Comment (bsdjhb) [reply sys/arm64/arm64/exception.S:132]: There are definitely some other places that check for hybrid in the kernel, but I don't know that it's widespread enough to warrant a new macro:

```
sys/arm64/include/asm.h:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/dev/pci/pci_user.c:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/kern/vfs_vnops.c:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/mips/mips/trap.c:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/riscv/riscv/copyinout.S:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/riscv/riscv/copyinout.S:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/riscv/riscv/copyinout.S:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/sys/systm.h:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/sys/systm.h:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
sys/sys/systm.h:#if __has_feature(capabilities) && !defined(__CHERI_PURE_CAPABILITY__)
```

File: sys/arm64/arm64/locore.S [hunk 3]
```diff
@@ -319,6 +319,15 @@ drop_to_el1:
 	/* Don't trap to EL2 for exceptions */
 	mov	x2, #CPTR_RES1
 	msr	cptr_el2, x2
+#if __has_feature(capabilities)
```
Comment (bsdjhb) [inline sys/arm64/arm64/locore.S:322]:
```suggestion

#if __has_feature(capabilities)
```

**Review by bsdjhb — APPROVED** [review]

File: sys/arm64/arm64/locore.S (hunk 3 above)
Comment (jrtc27) [reply sys/arm64/arm64/locore.S:322]: To be honest I don't know why we actually use cvbar_el2 and celr_el2 here, for a hybrid kernel we can just write to the non-capability part...

**Review by brettferdosi — APPROVED** [review]

**Comment by qwattash** [comment]
This is actually quite useful for the purecap morello kernel. FYI to avoid duplicating work, I have a WIP locore patch that does a similar thing with pointer registers.

**Comment by jrtc27** [comment]
> This is actually quite useful for the purecap morello kernel. FYI to avoid duplicating work, I have a WIP locore patch that does a similar thing with pointer registers.

Yeah, the talk of doing the purecap kernel is partly what motivated me to finally do this. I _nearly_ made use of PTR but figured I'd leave that to you as there'd surely still be some changes needed.

**Pull request merged**